This commit adds a simple check to see if curl is build against NSS.

apettinen opened the pull request:
Curl built against NSS has an effect on how cipher suites are handled in curl, i.e. NSS requires different syntax on cipher suites. I've only included the current cipher suite used by Apple, but the list could be longer.

I encountered this issue on:
`Linux our.reposado.server 2.6.32-642.3.1.el6.i686 #1 SMP Sun Jun 26 18:18:18 EDT 2016 i686 i686 i386 GNU/Linux`
curl version:
`curl 7.19.7 (i686-redhat-linux-gnu) libcurl/7.19.7 NSS/3.19.1 Basic ECC zlib/1.2.3 libidn/1.18 libssh2/1.4.2`

gregneagle commented:
See this discussion:
https://groups.google.com/d/topic/reposado/_fUl1n3uPq8/discussion

IOW, you could have addressed the issue for yourself with no code changes.

I'm not opposed to your changes; I wonder, though, where you got those specific values, and if they are always applicable when curl is built against NSS. I notice that you use:

`ciphers = rsa_aes_128_gcm_sha_256`

and the poster on the other thread used

`ciphers = "rsa_aes_128_sha,rsa_aes_256_sha,rsa_3des_sha"`

I'm not knowledgeable enough to know/say/decide which is better/more accurate/preferable.

My inclination, then, is to make no code changes at all and just rely on the current functionality, which allows the reposado admin to specify whichever curl ciphers they wish.

Reactions?

algolsigfrid commented:
Hi Greg,

I was the one who posted the configuration string on the discussion tread you linked.
To get to those particular settings I enumerated the ciphers supported by _swscan.apple.com_
and then checked the NSS documentation to match the ciphers with their naming scheme.

Time has passed and apple has changed the ciphers supported by _swscan.apple.com_ and today it supports only AES based ciphers, removing 3DES. The point being, that apple has effectively obsoleted the configuration I proposed. 

An administrator will need to keep checking and adjusting for any changes on the swscan.apple.com 

My two cents on this would be to actually remove the current hard coded settings

`'ciphers = HIGH,!ADH'`

In this way curl will use whatever ciphers it has available to it. Regardless of which crypto library was used to compile curl (NSS, OpenSSL, GnuTLS)
Any connection to _swscan.apple.com_ will only succeed if it uses the ciphers supported by that server.
On that regard, this limits which ciphers will be used. At this time only AES128 will work to connect to that server.
In addition, most current crypto libraries have better defaults disabling weaker ciphers. Just like any web browser. 
Finally, if the administrator still wants to hard code a cipher they can do so via the preferences file.

Thank you

gregneagle commented:
This sounds reasonable. I actually cannot remember why the ciphers were hard-coded in the first place. If there are no objections let's close this PR. If someone wants to open a new PR that removes the hard-coded ciphers I'll merge that, otherwise I'll try to remember to do it soon.

apettinen commented:
Yeah, I agree. I tested with automatic updating of the ciphers, but, IMHO, the cost/benefit ratio is not very good. For the gain, the process is overly complicated and adds unnecessary burden - and as @algolsigfrid pointed out, pretty much pointless. 

As I kinda started this whole thing, I think it's my duty to finish it. Will look into the code next week and do a PR. ok?

apettinen commented:
The commit is above, do you merge that or do I make a new PR?

I removed the part regarding ciphers and added equivalent of --tlsv1 on the curl_temp file. The adding of --tlsv1 is not strictly necessary.

I've tested this on our RHELs and it has been running in prod for good part of this week.

gregneagle commented on code/repo_sync:
```diff
@@ -294,10 +295,7 @@ def curl(url, destinationpath, onlyifnewer=False, etag=None, resume=False):
-        #print >> fileobj, 'output = "%s"' % tempdownloadpath
-
-        print >> fileobj, 'ciphers = HIGH,!ADH' # use only >=128 bit SSL
+        print >> fileobj, 'tlsv1'          # use only TLS 1.x
```
I assume you tested this on a few platforms and with curl compiled against NSS, etc?
(I thought was agreed it was safest to just leave all this stuff out...)

gregneagle replied:
I like the adding of tlsv1 conceptually -- just worried it will break some combos of OS version/curl and there's no easy way for the admin to override.

apettinen replied:
Well, I the oldest one I have available is RHEL6 with NSS. I don't have the ability to test on RHEL5 or older.

Adding tlsv1 only disables sslv2 and sslv3 (but allows tlsv1, 1.1 and 1.2). Apple does not support sslv2 or sslv3 anyway, as per:
```
╭  :~/
╰ ➤ openssl s_client -connect swscan.apple.com:443 -ssl3
CONNECTED(00000003)
write:errno=54

╭  :~/
╰ ➤ openssl s_client -connect swscan.apple.com:443 -ssl2
CONNECTED(00000003)
write:errno=54
```

apettinen replied:
Forgot to mention that --tlsv1 works on OS X/macOS 10.9.5, which is the oldest version I have access to.

gregneagle replied:
Quick testing seems to indicate this change doesn't explode with curl on 10.6, which the oldest thing reposado was ever supported on that I can still make run on hardware I have access to...

erikng commented:
Should be an easy test but I think so long as TLS 1.2 isn't the only allowed connection, it should work from 10.6 and up.

On Jan 19, 2017, at 9:08 AM, Greg Neagle <notifications@github.com<mailto:notifications@github.com>> wrote:


@gregneagle commented on this pull request.

________________________________

In code/repo_sync<https://github.com/wdas/reposado/pull/52>:

> @@ -294,10 +295,7 @@ def curl(url, destinationpath, onlyifnewer=False, etag=None, resume=False):
         print >> fileobj, 'fail'           # throw error if download fails
         print >> fileobj, 'dump-header -'  # dump headers to stdout
         print >> fileobj, 'speed-time = 30' # give up if too slow d/l
-        # next line does not work on windows
-        #print >> fileobj, 'output = "%s"' % tempdownloadpath
-
-        print >> fileobj, 'ciphers = HIGH,!ADH' # use only >=128 bit SSL
+        print >> fileobj, 'tlsv1'          # use only TLS 1.x


I like the adding of tlsv1 conceptually -- just worried it will break some combos of OS version/curl and there's no easy way for the admin to override.

-
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub<https://github.com/wdas/reposado/pull/52>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AFa_GGqx5S1-yUmEGiHQo5IgS9i8ou5Pks5rT5hsgaJpZM4JhB8i>.

apettinen commented:
@erikng yep, --tlsv1 allows TLS1, TLS1.1 and TLS1.2
